sm/plugins/tedge_apama_plugin/src/main.rs

```diff
@@ -0,0 +1,181 @@
+mod error;
+
+use crate::error::InternalError;
+use serde::Deserialize;
+use std::fs::{self, File};
+use std::os::unix::prelude::ExitStatusExt;
+use std::path::Path;
+use std::process::{Command, ExitStatus, Stdio};
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+struct ApamaCli {
+    #[structopt(subcommand)]
+    operation: PluginOp,
+}
+
+#[derive(StructOpt)]
+pub enum PluginOp {
+    /// List all the installed modules
+    List,
+
+    /// Install a module
+    Install {
+        module: String,
+        #[structopt(short = "v", long = "--module-version")]
+        version: Option<String>,
+        #[structopt(long = "--file")]
+        file_path: String,
+    },
+
+    /// Uninstall a module
+    Remove {
+        module: String,
+        #[structopt(short = "v", long = "--module-version")]
+        version: Option<String>,
+    },
+
+    /// Prepare a sequences of install/remove commands
+    Prepare,
+
+    /// Finalize a sequences of install/remove commands
+    Finalize,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum UpdateAction {
+    Install,
+    Remove,
+}
+#[derive(Debug, Deserialize)]
+struct SoftwareModuleUpdate {
+    pub action: UpdateAction,
+    pub name: String,
+    #[serde(default)]
+    pub version: Option<String>,
+    #[serde(default)]
+    pub path: Option<String>,
+}
+
+const APAMA_ENV_EXE: &str = "/opt/softwareag/Apama/bin/apama_env";
+const TEDGE_APAMA_PROJECT_DIR: &str = "/etc/tedge/apama/project";
+const TMP_APAMA_PROJECT_DIR: &str = "/tmp/tedge_apama_project";
+
+fn run(operation: PluginOp) -> Result<ExitStatus, InternalError> {
+    let success = ExitStatus::from_raw(0);
+
+    let tedge_env_exe_path = Path::new(APAMA_ENV_EXE);
+
+    if !tedge_env_exe_path.exists() {
+        return Err(InternalError::ApamaNotInstalled);
+    }
+
+    let tedge_apama_project_path = Path::new(TEDGE_APAMA_PROJECT_DIR);
+    let tmp_apama_project_path = Path::new(TMP_APAMA_PROJECT_DIR);
+
+    let status = match operation {
+        PluginOp::List => {
+            if tedge_apama_project_path.exists() {
+                println!("project\t")
+            }
+            success
+        }
+
+        PluginOp::Prepare => success,
+
+        PluginOp::Finalize => {
+            fs::remove_dir_all(tmp_apama_project_path)?;
+            success
+        }
+
+        PluginOp::Install {
+            module: _,
+            version: _,
+            file_path,
+        } => {
+            let archive_path = Path::new(&file_path);
+            let archive_file = File::open(&archive_path)?;
+
+            let mut archive = zip::ZipArchive::new(archive_file)?;
+
+            // TODO: Validate the zip to be valid apama project before extraction?
```

Comment:
Not sure if we need to do any explicit validation here or just rely on the apama service to detect any issues with the project. Anyway, I've asked the Apama team if we can do some "cheap" validations like checking for the presence of some mandatory files or something like that in the archive.